### PR TITLE
DEPR: deprecate passing redshift (`z`) as a keyword argument

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
This is the continuation (and revival) of @nstarman's work in #15859. See that PR for motivation.

~It is still a work in progress as I haven't applied my decorator anywhere yet. I'm opening as a draft to check for compilation and other installation errors on all platforms.~
Close #15859

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

TODO:
- [x] pass existing CI
- [x] add tests (I already wrote some, just need to port them to this PR)
- [x] add a docstring (and any documentation required)
- [x] apply decorator instead of making arguments positional only directly
- [x] edit changelog

### CODE REVIEW

**Comment by neutrinoceros** [comment]
@nstarman I just realized that we could skip the deprecation cycle for private methods. Do you agree ?

**Comment by nstarman** [comment]
Yes! Only public functions.

**Comment by neutrinoceros** [comment]
Almost there. I collapsed my branch after I made sure all lights were green. Before I undraft this, I'd like to give @nstarman priority for reviewing since this is based off his work and largely co-authored.

File: astropy/utils/tests/test_decorators.py
Comment (neutrinoceros) [inline astropy/utils/tests/test_decorators.py:917]:
```suggestion
                id="4 deprecations, 4 warns",
```

File: astropy/utils/tests/test_decorators.py
Comment (neutrinoceros) [inline astropy/utils/tests/test_decorators.py:924]:
```suggestion
                id="4 deprecations, 3 warns",
```

File: astropy/utils/tests/test_decorators.py
Comment (neutrinoceros) [inline astropy/utils/tests/test_decorators.py:931]:
```suggestion
                id="4 deprecations, 2 warns",
```

File: astropy/utils/tests/test_decorators.py
Comment (neutrinoceros) [inline astropy/utils/tests/test_decorators.py:938]:
```suggestion
                id="4 deprecations, 1 warn",
```

**Review by neutrinoceros — COMMENTED** [review]
Leaving myself a comment from my phone.

File: astropy/cosmology/flrw/base.py
Comment (nstarman) [inline astropy/cosmology/flrw/base.py:88]: Haven't looked through the C-code yet, but does `@deprecated_keywords` map to the positions? As in, can this be made positional-only now? And then we remove the decorator later, to complete the deprecation?

Comment (neutrinoceros) [reply astropy/cosmology/flrw/base.py:88]: It doesn't. Maybe it's possible but I don't know the C API well enough to see how it could be implemented.

**Comment by mhvk** [comment]
I'm on holidays this week, but FWIW happy with the general idea.

File: astropy/cosmology/flrw/base.py
Comment (neutrinoceros) [reply astropy/cosmology/flrw/base.py:88]: Actually I *do* think it's feasible, but it's a trade-of for where we put complexity: it would certainly make the C extension more complex and I'm already pushing my own limit there so I'm not confident that I could build this feature without hurting robustness and readability. How about I give it a try as a follow up task ?

Comment (nstarman) [reply astropy/cosmology/flrw/base.py:88]: I think that's totally reasonable!  And thanks again for taking the lead on this!
Having the parameters be positional-only at the undecorated Python function level is an improvement, but to the user there is no difference, so the PR as written is great 👍.

File: astropy/utils/decorators.py
Comment (nstarman) [inline astropy/utils/decorators.py:638]: A quick test for this would be great.

File: astropy/utils/decorators.py
Comment (nstarman) [inline astropy/utils/decorators.py:40]: I think you can do 
```suggestion
_F = TypeVar("_F", bound=Callable[..., Any])
```

**Review by nstarman — APPROVED** [review]
Thanks @neutrinoceros! This looks great. I would love it if someone else also reviewed the C-code, as I'm much less conversant in C-to-Python.

File: astropy/utils/decorators.py
Comment (pllim) [inline astropy/utils/decorators.py:17]: I am not familiar with cosmology code base, so please explain to me why deprecating `z` in cosmology needs C extension in `utils`. Thanks!

Comment (pllim) [reply astropy/utils/decorators.py:17]: Ah, I just saw https://github.com/astropy/astropy/pull/15859#issuecomment-2180747472 . Performance boost is always welcome but I am afraid this C code would also make the decorator hard to maintain going forward. I will put this on queue to review.

Comment (pllim) [reply astropy/utils/decorators.py:17]: Oh, so this boost is only for a very specific case. And I feel like the new function (ignoring performance) duplicated something `deprecated_renamed_argument` can already do. I need to ponder.

Even if we decide to keep this as-is, license for https://pypi.org/project/positional-defaults/ would need to be included in https://github.com/astropy/astropy/tree/main/licenses

Comment (pllim) [reply astropy/utils/decorators.py:17]: Also the new function is technically new API, so might also need a change log for `utils`.

Comment (neutrinoceros) [reply astropy/utils/decorators.py:17]: > And I feel like the new function (ignoring performance) duplicated something `deprecated_renamed_argument` can already do.

I don't think that's the case. `deprecated_renamed_argument` only works if a keyword argument is still allowed, albeit under a new name, while `deprecated_keyword` forces the user to stop using the argument as keyword.

> I am afraid this C code would also make the decorator hard to maintain going forward
> Also the new function is technically new API

Good points. We don't actually need to make it public API. Maybe we can keep it private for now and see where it goes ? If it's not used at all after this deprecation cycle is concluded we can just get rid of this C extension altogether ?

> Even if we decide to keep this as-is, license for [pypi.org/project/positional-defaults](https://pypi.org/project/positional-defaults/) would need to be included in [main/licenses](https://github.com/astropy/astropy/tree/main/licenses?rgh-link-date=2024-06-24T19%3A28%3A54Z)

Thanks !

File: astropy/cosmology/flrw/base.py
Comment (neutrinoceros) [reply astropy/cosmology/flrw/base.py:88]: dully noted !

File: astropy/utils/decorators.py
Comment (neutrinoceros) [reply astropy/utils/decorators.py:638]: actually this is vestigial of `positional_defaults`, which for some reason had more complexity in that the equivalent function was overloaded. I simplified it so I might as well drop this branch too.

**Comment by mhvk** [comment]
Definitely from the sidelines, but this seems an inordinate amount of extra code for what essentially is beautification of the code. After all, nothing is stopping people now from using a positional argument; all this does is stop people from using `z=...`. Is this worth it?

**Comment by neutrinoceros** [comment]
@nstarman do you want to reply here ?

**Comment by nstarman** [comment]
> Is this worth it?

Yes! There are a few motivating factors behind this change:

1. Clarity. There are more concepts in math/physics than there are symbols. `z` is the common symbol used for redshift in Cosmology, but also for complex numbers, and now that ML has permeated cosmology also for latent spaces. It's often clear in context, both mathematically and in code, but it's worthwhile asking "what should I name my function argument?". `redshift` is very clearly a better variable / argument name as it's much less ambiguous, but it's also 8x more verbose. `redshift` also isn't perfect because cosmology functions are interested in the cosmological redshift, not doppler redshift. So `cosmo_redshift` would be a yet-more-clear function name. But that is 14x more verbose than `z`. Thankfully positional-only arguments mitigate this problem. Let's see PEP 570:

> When designing APIs (application programming interfaces), library authors try to ensure correct and intended usage of an API. **Without the ability to specify which parameters are positional-only, library authors must be careful when choosing appropriate parameter names.** This care must be taken even for required parameters or when the parameters have no external semantic meaning for callers of the API.

All of the functions that are being updated take the redshift as their first (or first 2) arguments. The meaning of the function is contextually clear **even without the variable names** (which are still visible even when positional-only and further clarify meaning). So given that `z` is only a decent variable name choice and that there are a large number of appropriate and clearer variable names, providing 100% clarity by making the argument positional-only is simply better API.

2. Downstream Impact: Minimal because almost everyone passes it in primarily by position. For the less-common cases there's a deprecation period. Any code just needs to be changed from `(z=x, ...)` to  (x)`. I took a look at GH and didn't find any counterexamples to this easy transition.

3. Interoperability with other libraries. From point 1 I discussed how there are reasonable (and possibly better) variable names than `z`. Libraries that build upon `astropy.cosmology` even think so. `SkyPy` uses `redshift` as its variable name of choice. Making this argument positional-only is a better API choice for connecting to other libraries. They can name a variable whatever they want (or better yet make it positional) and people (once they are familiar with the change in Astropy) will not have any mental dissonance about `redshift` versus `z` because positional-only arguments are positional, not named. Per PEP 570:

> Positional-only parameters give more control to library authors to better express the intended usage of an API and allows the API to evolve in a safe, backward-compatible way.

Yes plz.

Last, as more a general philosophical point. These methods were written before positional-only arguments were a thing in Python. So when written they had to be positional-or-keyword. Once PEP 570 was implemented it is good to look at the code and ask where  PEP 570 is relevant and can improve our API. This is one such place. We can and should improve the API in ways big and small. The reason this PR is heavy is because we are also dedicated to long-term support and smooth transitions. This PR could be very lightweight if we didn't offer a performant deprecation period!

As a possible alternative, we could move the decorator into `cosmology` rather than `utils`. It can be temporary, added only for the length of the deprecation period.

File: astropy/utils/decorators.py
Comment (nstarman) [reply astropy/utils/decorators.py:17]: > Good points. We don't actually need to make it public API

Agreed. Let's make it private. If @mhvk or @pllim prefers, we can also move it into `cosmology`. Though looking around `Astropy` I would recommend other maintainers avail themselves of this useful utility and improve API by making some required arguments positional-only.

**Comment by pllim** [comment]
Re: decorator -- If you think the performance is irrelevant outside of this narrow set of code, then it is best to keep it super private and then remove as @nstarman suggested.

Otherwise, I fear it could have package-wide implication. And I think the decorator speed-up is best done as a separate PR. And should be used more widely in other subpackages as well to be consistent.

**Comment by neutrinoceros** [comment]
Ok, let's make the decorator a private asset of `astropy.cosmology`. It doesn't even prevent promoting it to public API later if it ever grows in usage.

> And I think the decorator speed-up is best done as a separate PR.

To be clear, are you suggesting I re-implement the decorator in Python only for the sake of delaying the more painful review ?
I don't mind re-implementing it in Python (I actually already have a mostly drop-in Python replacement), I'm just not sure that it actually helps; tests are already demonstrating how it works on the Python side. I don't mind going the longer way if requested, I just want to make sure that we all understand it is more work for all parties.

> And should be used more widely in other subpackages as well to be consistent.

Sorry I don't get what you mean here.

**Comment by pllim** [comment]
> To be clear, are you suggesting I re-implement the decorator in Python only for the sake of delaying the more painful review ?

> Sorry I don't get what you mean here.

No, I mean, if you want to keep the C-extension and have it exposed via `utils`. We should also see if there are other similar deprecations that can use it as well.

> let's make the decorator a private asset of `astropy.cosmology`

I am uneasy with this but it would be much less controversial.

Maybe we need to step back and rethink this whole thing. Let's look at this:

```python
@deprecated_keywords("z", since="7.0")
def scale_factor(self, z):
    # ...
```

The signature has always suggests `z` is positional. It is not our problem that Python lets people also pass it in as `z=`. We never advertised it as such, did we? How many people out there in reality is actually passing in `z=`?

Can we just do this and be done with it?

```python
def scale_factor(self, z, /):
    # ...
```

In the change log, we can say like "ooops you been using it in a way we didn't mean to support, you cannot do that anymore in 7.0" and that is probably okay because it is a major release and breaking change is allowed.

**Comment by neutrinoceros** [comment]
In the absence of a clear deprecation path, biting the bullet would make sense to me, but 
now that there's one, I don't understand why we would risk antagonizing our users when we don't have to.

**Comment by pllim** [comment]
How many users are we antagonizing? If it is just one or two, I will take that risk. I will even open a PR in their repos.

**Comment by pllim** [comment]
Also if performance is really such a concern to the point you need C-extension for a decorator, biting bullet with `/` would be the most performant.

**Comment by neutrinoceros** [comment]
> How many users are we antagonizing? If it is just one or two, I will take that risk. I will even open a PR in their repos.

No idea. Which is precisely why I'd err on the side of caution and avoid taking a risk. I also assume that a majority of our users wouldn't provide feedback or make their code public, so I would rather not create a precedent where we _could_ have avoided immediate breakage for them and didn't.

**Comment by mhvk** [comment]
Again from the sidelines - while I would probably still not bother myself, the arguments for making it positional-only (eventually) are reasonable and now that one has done the work, it is perhaps a bit silly to not implement this. I think making it a cosmology-internal function is the way to go - if we need it elsewhere, we can still move it, but this way it is clear the whole stuff can go once the deprecation period is over. Indeed, would be good to open an issue now for 8.0(?) to remind us to remove all of it.

**Comment by neutrinoceros** [comment]
Here you go: https://github.com/astropy/astropy/issues/16614

**Comment by nstarman** [comment]
> How many users are we antagonizing? If it is just one or two, I will take that risk. I will even open a PR in their repos.

Surprisingly a lot. https://github.com/search?q=distance%28z%3D&type=code is a quick search on just a small slice of our API.

File: astropy/utils/decorators.py
Comment (pllim) [inline astropy/utils/decorators.py:1]: Please revert this change.

File: astropy/utils/setup_package.py
Comment (pllim) [inline astropy/utils/setup_package.py:1]: Please revert this change.

**Review by pllim — COMMENTED** [review]
Since I have been outvoted, I'll let Nathaniel or Marten to have one last look and merge. Just asking you to remove the diff from `utils`. Thanks!

**Comment by neutrinoceros** [comment]
done. I've also squashed my commits together (but left @nstarman's alone)

File: astropy/cosmology/_utils.py
Comment (pllim) [inline astropy/cosmology/_utils.py:1]: I just want to make @adrn aware of yet another utils module. (Inside joke...)

**Comment by neutrinoceros** [comment]
@nstarman this is ready to go !

**Comment by nstarman** [comment]
Yes! I hadn't notice that you'd moved the C code into `cosmology`.
Let's get this in!

**Comment by eerovaher** [comment]
I haven't thought about using positional-only function arguments in `coordinates` and I don't plan to think about that in the near future either. But maybe @adrn has ideas about this.

File: astropy/cosmology/flrw/base.py
Comment (neutrinoceros) [reply astropy/cosmology/flrw/base.py:92]: In hindsight and given the complaints this PR got for introducing a hard-to-maintain C extension, I'm not sure it's worth pushing it any further without a clear need to.

**Pull request merged**
